File: 2term/jvmWorld/2lab_kotlin_btree/src/main/kotlin/Point2D.kt
import kotlin.math.pow
import kotlin.math.sqrt
import kotlin.random.Random
import java.math.RoundingMode

class Point2D(val x: Double, val y: Double) : CustomTypeTrait {

    override val typeName: String
        get() = "Point2D"

    override val typeExampleRepresentation: String
        get() = "0.5|7.8"

    override fun createRandomInstance(): Point2D {
        val x = (Random.nextDouble() * 20 - 10).toBigDecimal().setScale(2, RoundingMode.HALF_UP).toDouble()
        val y = (Random.nextDouble() * 20 - 10).toBigDecimal().setScale(2, RoundingMode.HALF_UP).toDouble()
        return Point2D(x, y)
    }

    override fun toString(): String {
        return "X: $x, Y: $y"
    }

    override fun serializeToString(): String {
        return "$x|$y"
    }

    override fun deserializeFromString(str: String): Point2D? {
        val parts = str.split('|')
        return if (parts.size == 2) {
            try {
                val x = parts[0].toDouble()
                val y = parts[1].toDouble()
                Point2D(x, y)
            } catch (e: NumberFormatException) {
                null
            }
        } else {
            null
        }
    }

    val distanceFromOrigin: Double
        get() = sqrt(x.pow(2) + y.pow(2))

    override fun compare(other: CustomTypeTrait): Int {
        return when (other) {
            is Point2D -> {
                val distanceA = this.distanceFromOrigin
                val distanceB = other.distanceFromOrigin
                distanceA.compareTo(distanceB)
            }
            else -> throw IllegalArgumentException("Cannot compare Point with a different type")
        }
    }
}

File: 2term/jvmWorld/1lab_scala_binaryTree/src/main/scala/Point2D.scala
import scala.math.BigDecimal.RoundingMode
import scala.math.{pow, sqrt}
import scala.util.{Random, Try}


class Point2D(val x: Double, val y: Double) extends CustomTypeTrait {

  override def typeName: String = "2DPoint"

  override def typeExampleRepresentation: String = "0.5|7.8"

  override def createRandomInstance(): this.type = {
    val r = new Random()
    val x = BigDecimal(r.nextDouble() * 20 - 10).setScale(2, RoundingMode.HALF_UP).toDouble
    val y = BigDecimal(r.nextDouble() * 20 - 10).setScale(2, RoundingMode.HALF_UP).toDouble
    new Point2D(x, y).asInstanceOf[this.type]
  }

  override def toString: String = s"X: $x, Y: $y"

  override def serializeToString: String = {
    s"$x|$y"
  }

  override def deserializeFromString(str: String): Option[this.type] = {
    str.split('|') match {
      case Array(xStr, yStr) =>
        try {
          val x = xStr.toDouble
          val y = yStr.toDouble
          val instance = new Point2D(x, y)
          Some(instance.asInstanceOf[this.type])
        } catch {
          case _: NumberFormatException => None
        }
      case _ => None
    }
  }

  def distanceFromOrigin: Double = sqrt(pow(x, 2) + pow(y, 2))

  override def compare(other: CustomTypeTrait): Int = other match {
    case point: Point2D =>
      val distanceA = this.distanceFromOrigin
      val distanceB = point.distanceFromOrigin
      distanceA.compare(distanceB)
    case _ =>
      throw new IllegalArgumentException("Cannot compare Point with a different type")
  }
}
